src/main/scala/org/broadinstitute/hail/stats/LinearMixedModel.scala

```diff
@@ -0,0 +1,303 @@
+    val Ty = y :* sqrtInvD
+    val TyTy = Ty dot Ty
+    val TCTy = TC.t * Ty
+    val TCTC = TC.t * TC
+    val b = TCTC \ TCTy
+    val s2 = (TyTy - (TCTy dot b)) / (if (useML) n else n - C.cols)
+
+    DiagLMM(b, s2, math.log(s2), delta, maxLogLkhd, gridLogLkhd, sqrtInvD, TC, Ty, TyTy, useML)
+  }
+
+  def fitDelta(C: DenseMatrix[Double], y: DenseVector[Double], S: DenseVector[Double], useML: Boolean): (Double, Option[Double], Option[IndexedSeq[(Double, Double)]]) = {
+
+    val n = y.length
+    val c = C.cols
+    val shift =
+      if (useML)
+        n + n * math.log(2 * math.Pi)
+      else
+        n + (n - c) * math.log(2 * math.Pi) - logdet(C.t * C)._2
+
+    object LogLkhdML extends UnivariateFunction {
+      def value(delta: Double): Double = {
+        val D = S + delta
+        val dy = y :/ D
+        val ydy = y dot dy
+        val Cdy = C.t * dy
+        val CdC = C.t * (C(::, *) :/ D)
+        val b = CdC \ Cdy
+        val r = ydy - (Cdy dot b)
+
+        -.5 * (sum(breeze.numerics.log(D)) + n * math.log(r) + shift)
+      }
+    }
+
+    object LogLkhdREML extends UnivariateFunction {
+      def value(delta: Double): Double = {
+        val D = S + delta
+        val dy = y :/ D
+        val ydy = y dot dy
+        val Cdy = C.t * dy
+        val CdC = C.t * (C(::, *) :/ D)
+        val b = CdC \ Cdy
+        val r = ydy - (Cdy dot b)
+
+        -.5 * (sum(breeze.numerics.log(D)) + (n - c) * math.log(r) + logdet(CdC)._2 + shift)
+      }
+    }
+
+    val logmin = -10
+    val logmax = 10
+    val pointsPerUnit = 100 // number of points per unit of log space
+
+    val grid = (logmin * pointsPerUnit to logmax * pointsPerUnit).map(_.toDouble / pointsPerUnit) // avoids rounding of (logmin to logmax by logres)
+
+    val gridLogLkhd = if (useML)
+      grid.map(logDelta => (logDelta, LogLkhdML.value(math.exp(logDelta))))
+    else
+      grid.map(logDelta => (logDelta, LogLkhdREML.value(math.exp(logDelta))))
+
+    val header = "logDelta\tlogLkhd"
+    val gridValsString = gridLogLkhd.map{ case (d, nll) => s"$d\t$nll" }.mkString("\n")
+    log.info(s"\nlmmreg: table of delta\n$header\n$gridValsString\n")
+
+    val approxLogDelta = gridLogLkhd.maxBy(_._2)._1
+
+    if (approxLogDelta == logmin)
+      fatal(s"lmmreg: failed to fit delta: ${if (useML) "ML" else "REML"} realized at delta lower search boundary e^$logmin = ${math.exp(logmin)}, indicating negligible enviromental component of variance. The model is likely ill-specified.")
+    else if (approxLogDelta == logmax)
+      fatal(s"lmmreg: failed to fit delta: ${if (useML) "ML" else "REML"} realized at delta upper search boundary e^$logmax = ${math.exp(logmax)}, indicating negligible genetic component of variance. Standard linear regression may be more appropriate.")
+
+    val searchInterval = new SearchInterval(-10d, 10d, approxLogDelta)
+    val goal = GoalType.MAXIMIZE
+    val objectiveFunction = new UnivariateObjectiveFunction(if (useML) LogLkhdML else LogLkhdREML)
```

Comment:
In fact, I'd just make an implicit conversion from `(Double) => Double` to `UnivariantFunction`.  That's all it is, anyway.